core/adt/array.hpp

```diff
@@ -6,61 +6,86 @@
 
-  /**
-   * Container for storing values preserving their order of insertion.
-   * Implementation is based on Array Mapped Trie.
-   */
+  template <typename Value>
   struct Array {
-    using Value = storage::ipfs::IpfsDatastore::Value;
-    using IndexedVisitor =
-        std::function<outcome::result<void>(uint64_t, const Value &)>;
-    using Visitor = std::function<outcome::result<void>(const Value &)>;
-
-    explicit Array(const std::shared_ptr<storage::ipfs::IpfsDatastore> &store);
-    Array(const std::shared_ptr<storage::ipfs::IpfsDatastore> &store,
-          const CID &root);
-
-    /**
-     * Apply changes to storage
-     * @return root CID
-     */
-    outcome::result<CID> flush();
-
-    /// Appends a value. Does not change the store
-    outcome::result<void> append(gsl::span<const uint8_t> value);
-
-    /**
-     * Appends a value which can be CBOR-encoded.
-     * Does not change the store immediately
-     * @tparam T value type that supports CBOR encoding
-     * @param value to be marshalled
-     * @return operation result
-     */
-    template <typename T>
-    outcome::result<void> appendCbor(const T &value) {
-      OUTCOME_TRY(bytes, codec::cbor::encode(value));
-      return append(bytes);
+    using Key = uint64_t;
+    using Visitor = std::function<outcome::result<void>(Key, const Value &)>;
+
+    Array() : amt{nullptr} {}
+
+    // NOLINTNEXTLINE(google-explicit-constructor)
+    Array(const CID &root) : amt{nullptr, root} {}
+
+    void load(std::shared_ptr<Ipld> ipld) {
+      amt.setIpld(ipld);
+    }
+
+    outcome::result<boost::optional<Value>> tryGet(Key key) {
+      auto maybe = get(key);
+      if (!maybe) {
+        if (maybe.error() != storage::amt::AmtError::NOT_FOUND) {
+          return maybe.error();
+        }
+        return boost::none;
+      }
+      return maybe.value();
+    }
+
+    outcome::result<bool> has(Key key) {
+      return amt.contains(key);
+    }
+
+    outcome::result<Value> get(Key key) {
+      return amt.getCbor<Value>(key);
+    }
+
+    outcome::result<void> set(Key key, const Value &value) {
+      return amt.setCbor(key, value);
+    }
+
+    outcome::result<void> remove(Key key) {
+      return amt.remove(key);
     }
 
-    /// Iterate over stored elements acquiring their index
-    outcome::result<void> visit(const IndexedVisitor &visitor);
+    outcome::result<void> flush() {
+      OUTCOME_TRY(amt.flush());
+      return outcome::success();
+    }
 
-    /// Iterate over stored elements
-    outcome::result<void> visit(const Visitor &visitor);
+    outcome::result<void> visit(const Visitor &visitor) {
+      return amt.visit([&](auto key, auto &value) -> outcome::result<void> {
+        OUTCOME_TRY(value2, codec::cbor::decode<Value>(value));
+        return visitor(key, value2);
+      });
+    }
 
-   private:
-    std::shared_ptr<storage::ipfs::IpfsDatastore> store_;
-    storage::amt::Amt amt_;
+    storage::amt::Amt amt;
```

Comment:
why did you decide to expose `amt` field?

Reply:
Defining template cbor encoding/decoding methods as friends will fail, because they use default template parameter (`typename = enable_if_t<...>`).

Also this class is a wrapper, so amt should be accessible for those who need it.